session/session.go

```diff
@@ -2604,16 +2604,194 @@ func (s *session) Auth(user *auth.UserIdentity, authentication, salt []byte) err
 	if err != nil {
 		return privileges.ErrAccessDenied.FastGenByArgs(user.Username, user.Hostname, hasPassword)
 	}
-	if err = pm.ConnectionVerification(user, authUser.Username, authUser.Hostname, authentication, salt, s.sessionVars.TLSConnectionState); err != nil {
+	//Check whether continuous login failure is enabled to lock the account.
+	//If enabled, determine whether to unlock the account and notify TiDB to update the cache
+	enableAutoLock := pm.IsAccountAutoLockEnabled(authUser.Username, authUser.Hostname)
+	if enableAutoLock {
+		passwordLockingJSON, err := pm.VerifyAccountAutoLock(authUser.Username, authUser.Hostname)
+		if err != nil {
+			return err
+		}
+		if passwordLockingJSON != "" {
+			if err = s.passwordLocking(authUser.Username, authUser.Hostname, passwordLockingJSON); err != nil {
+				return err
+			}
+			err = domain.GetDomain(s).NotifyUpdatePrivilege()
+			if err != nil {
+				return err
+			}
+		}
+	}
+	err = pm.ConnectionVerification(user, authUser.Username, authUser.Hostname, authentication, salt, s.sessionVars.TLSConnectionState)
+	if err != nil {
+		//when user enables the account locking function for consecutive login failures,
+		//the system updates the login failure count and determines whether to lock the account when authentication fails
+		if userPasswordFailedErr, ok := err.(*privileges.ErrUserPasswordFailed); ok {
+			if enableAutoLock {
+				if trackingErr := authFailedTracking(s, authUser.Username, authUser.Hostname); trackingErr != nil {
+					return trackingErr
+				}
+			}
+			return userPasswordFailedErr.Err
+		}
 		return err
 	}
+	if enableAutoLock {
+		return authSuccessClearCount(s, authUser.Username, authUser.Hostname)
+	}
+	pm.AuthSuccess(authUser.Username, authUser.Hostname)
 	user.AuthUsername = authUser.Username
 	user.AuthHostname = authUser.Hostname
 	s.sessionVars.User = user
 	s.sessionVars.ActiveRoles = pm.GetDefaultRoles(user.AuthUsername, user.AuthHostname)
 	return nil
 }
 
+func authSuccessClearCount(s *session, user string, host string) error {
+	passwordLocking, getErr := getFailedLoginCount(s, user, host)
+	if getErr != nil {
+		if rollBackErr := failedLoginTrackingRollback(s); rollBackErr != nil {
+			return rollBackErr
+		}
+		return getErr
+	}
+	if passwordLocking.AutoAccountLocked {
+		if passwordLocking.PasswordLockTimeDays == -1 {
```

Comment:
Why we don't have a `ROLLBACK`(`failedLoginTrackingRollback`) here?

Reply:
for unlock or rollback begin

Reply:
The login failed session ended, but for safety it should be added

Reply:
done